core/src/main/java/com/yandex/yoctodb/util/mutable/impl/FoldedByteArrayIndexedList.java

```diff
@@ -0,0 +1,153 @@
+package com.yandex.yoctodb.util.mutable.impl;
+
+import com.google.common.primitives.Ints;
+import com.google.common.primitives.Longs;
+import com.yandex.yoctodb.util.OutputStreamWritable;
+import com.yandex.yoctodb.util.UnsignedByteArray;
+import com.yandex.yoctodb.util.mutable.ByteArrayIndexedList;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+final public class FoldedByteArrayIndexedList implements ByteArrayIndexedList {
+    @NotNull
+    private final List<UnsignedByteArray> elements;
+    private final List<UnsignedByteArray> uniqueElements;
+    private final List<Long> offsets;
+    private final List<Integer> offsetIndexes;
+    private final Map<UnsignedByteArray, Long> valueOffset;
+    private final int sizeOfIndexOffsetValue; // how many bites
+
+
+    public FoldedByteArrayIndexedList(
+            @NotNull final List<UnsignedByteArray> elements) {
+        this.elements = elements;
+        this.offsetIndexes = new ArrayList<>();
+        this.valueOffset = new HashMap<>();
+        this.offsets = new ArrayList<>();
+        this.uniqueElements = new ArrayList<>();
+
+        long elementOffset = 0;
+        for (int docId = 0; docId < elements.size(); docId++) {
+            UnsignedByteArray elem = elements.get(docId);
+            if (!valueOffset.containsKey(elements.get(docId))) {
+                valueOffset.putIfAbsent(elements.get(docId), elementOffset);
+                offsets.add(elementOffset);
+                elementOffset += elem.getSizeInBytes();
+                uniqueElements.add(elem);
+            }
+            offsetIndexes.add(offsets.indexOf(valueOffset.get(elem)));
+        }
+
+        offsets.add(elementOffset);
+
+        // analyze result of offsets
+        int offsetCount = offsets.size();
+        if (offsetCount <= 255) { // one byte 2^8 - 1 = 255
+            sizeOfIndexOffsetValue = 1;
+        } else if (offsetCount <= 65535) {  // to  2^16 - 1 = 65535
+            sizeOfIndexOffsetValue = 2;
+        } else {
+            sizeOfIndexOffsetValue = 4;
+        }
+
+    }
+
+    @Override
+    public long getSizeInBytes() {
+        long elementSize = 0;
+        for (UnsignedByteArray element : uniqueElements) {
+            elementSize += element.length();
+        }
+
+        return 4L + // Element count in bytes
+                4L + // offsets count in bytes
+                sizeOfIndexOffsetValue *
+                        elements.size() + // indexes offsets in bytes
+                8L * offsets.size() + // offsets array size in bytes
```

Comment:
They also hide in various methods for size calculation. Consider using Byte/Short/Integer/Long `BYTES` and `SIZE` constants